msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java

```diff
@@ -1064,51 +1060,181 @@ protected void acquireTokenSilent(@NonNull final String[] scopes,
+                    multiTenantAccount.getTenantId()
             );
+        } else if (authority instanceof AzureActiveDirectoryAuthority) {
+            final AzureActiveDirectoryAuthority aadAuthority = (AzureActiveDirectoryAuthority) authority;
 
-            ApiDispatcher.submitSilent(silentTokenCommand);
-        } catch (final BaseException exception) {
-            callback.onError(exception);
+            // Although the below call implies the returned value will be a tenantId, this isn't
+            // strictly true: it can also be an alias such as 'common', 'consumers', or
+            // 'organizations'. Additionally, if the developer has used a named tenant it could end
+            // up being something like <tenant_name>.onmicrosoft.com
+            //
+            // If the tenant is a GUID, we need to choose the account/profile that corresponds to it
+            //
+            // If the tenant is named 'common', 'consumers', or 'organizations' we need to use the
+            // home account
+            //
+            // If the account is named like <tenant_name>.onmicrosoft.com we need to query the OpenId
+            // Provider Configuration Metadata in order to get the tenant id. Once we have the
+            // tenant id, we must then select the appropriate home or profile.
+            final String tenantIdNameOrAlias = aadAuthority.getAudience().getTenantId();
+
+            // The AccountRecord we'll use to request a token...
+            final AccountRecord accountRecord = new AccountRecord();
+            accountRecord.setEnvironment(multiTenantAccount.getEnvironment());
+            accountRecord.setHomeAccountId(multiTenantAccount.getHomeAccountId());
+
+            if (isHomeTenantEquivalent(tenantIdNameOrAlias)
+                    || isAccountHomeTenant(multiTenantAccount.getClaims(), tenantIdNameOrAlias)) {
+                // Use home...
+                validateClaimsExistForTenant(tenantIdNameOrAlias, multiTenantAccount);
+
+                accountRecord.setLocalAccountId(multiTenantAccount.getId());
+                accountRecord.setUsername(multiTenantAccount.getUsername());
+
+                return accountRecord;
+            } else if (null != multiTenantAccount.getTenantProfiles().get(tenantIdNameOrAlias)) {
+                // Use this guest...
+                final ITenantProfile profileForRequest =
+                        multiTenantAccount
+                                .getTenantProfiles()
+                                .get(tenantIdNameOrAlias);
+
+                validateClaimsExistForTenant(tenantIdNameOrAlias, profileForRequest);
+
+                accountRecord.setLocalAccountId(profileForRequest.getId());
+                accountRecord.setUsername(profileForRequest.getUsername());
+
+                return accountRecord;
+            } else {
+                // We need to query metadata...
+                final OpenIdProviderConfigurationClient client =
```

Comment:
nit: suggest extracting this into a separate function.

Reply:
Resolved in 09c2c72fabfa2da4d71e79e22d402598d4db1410